core/drivers/crypto/hisilicon/hisi_qm.c

```diff
@@ -143,49 +148,113 @@ static void qm_db(struct hisi_qm *qm, uint16_t qn, uint8_t cmd, uint16_t index,
 	io_write64(qm->io_base + QM_DOORBELL_SQ_CQ_BASE, doorbell);
 }
 
+static void qm_mb_write(struct hisi_qm *qm, struct qm_mailbox *mb)
+{
+	vaddr_t dst = qm->io_base + QM_MAILBOX_BASE;
+
+	write_64bit_pair(dst, mb->x[0], mb->x[1]);
+}
+
+static void qm_mb_read(struct hisi_qm *qm, struct qm_mailbox *mb)
+{
+	vaddr_t mb_base = qm->io_base + QM_MAILBOX_BASE;
+
+	write_64bit_pair((vaddr_t)mb, mb_base, mb_base + U64_OFFSET);
+}
+
 static enum hisi_drv_status qm_wait_mb_ready(struct hisi_qm *qm)
 {
-	uint32_t val = 0;
+	struct qm_mailbox mb = { };
+	uint32_t i = 0;
 
-	/* Return 0 mailbox ready, HISI_QM_DRVCRYPT_ETMOUT hardware timeout */
-	if (IO_READ32_POLL_TIMEOUT(qm->io_base + QM_MAILBOX_BASE, val,
-				   !(val & QM_MB_BUSY_BIT), POLL_PERIOD,
-				   POLL_TIMEOUT)) {
-		return HISI_QM_DRVCRYPT_ETMOUT;
+	while (i++ < QM_MB_WAIT_READY_CNT) {
+		/* 128 bits should be read from hardware at one time*/
+		qm_mb_read(qm, &mb);
+		if (!(mb->w0 & QM_MB_BUSY_BIT))
+			return HISI_QM_DRVCRYPT_NO_ERR;
+		udelay(QM_MB_WAIT_PERIOD);
+	}
+
+	EMSG("QM mailbox is busy to start!");
+
+	return HISI_QM_DRVCRYPT_EBUSY;
+}
+
+static enum hisi_drv_status qm_wait_mb_finish(struct hisi_qm *qm,
+					      struct qm_mailbox *mb)
+{
+	uint32_t i = 0;
+
+	while (++i) {
```

Comment:
`++i` is always going to be true, with the check below. I'd prefer `for (i = 0; ; i++)` or `while (true)` and increasing `i` somewhere else in the loop, preferably not in the `if` below.

Reply:
solved